src/accountsdb/buffer_pool.zig

```diff
@@ -0,0 +1,1074 @@
+        file: std.fs.File,
+        file_id: FileId,
+        /// inclusive
+        range_start: FileOffset,
+        /// exclusive
+        range_end: FileOffset,
+    ) !CachedRead {
+        const indices = try self.makeindices(file_id, allocator, range_start, range_end);
+        errdefer allocator.free(indices);
+
+        var n_invalid: u32 = 0;
+        for (indices) |idx| {
+            if (idx == INVALID_FRAME) {
+                n_invalid += 1;
+            } else {
+                // we don't want to evict these in the next loop
+                self.eviction_lfu.insert(self.frames_metadata, idx);
+                if (!self.frames_metadata.rc[idx].acquire()) {
+                    // frame has no handles, but memory is still valid
+                    self.frames_metadata.rc[idx].reset(); // dumb
+                }
+            }
+        }
+
+        // fill in invalid frames with file data, replacing invalid frames with
+        // fresh ones.
+        for (0.., indices) |i, *idx| {
+            if (idx.* != INVALID_FRAME) continue;
+            // INVALID_FRAME => not found, read fresh and populate
+
+            const file_offset: FileOffset = @intCast((i * FRAME_SIZE) +
+                (range_start - range_start % FRAME_SIZE));
+            if (file_offset % FRAME_SIZE != 0) unreachable;
+
+            if (idx.* != INVALID_FRAME) continue;
+            idx.* = blk: while (true) {
+                if (self.free_list.popOrNull()) |free_idx| {
+                    break :blk free_idx;
+                } else {
+                    self.evictUnusedFrame();
+                }
+            };
+
+            const bytes_read = try file.pread(&self.frames[idx.*], file_offset);
+            if (bytes_read > FRAME_SIZE) unreachable;
+            self.populateNew(idx.*, file_id, file_offset, @intCast(bytes_read));
+        }
+
+        return CachedRead{
+            .bp = self,
+            .indices = indices,
+            .start_offset = @intCast(range_start % FRAME_SIZE),
+            .end_offset = @intCast(((range_end - 1) % FRAME_SIZE) + 1),
+        };
+    }
+};
+
+/// TODO: atomics on all index accesses
+pub const FramesMetadata = struct {
+    pub const InQueue = enum(u2) { none, small, main, ghost };
+
+    /// ref count for the frame. For frames that are currently being used elsewhere.
+    rc: []sig.sync.ReferenceCounter,
+
+    /// effectively the inverse of BufferPool.FrameMap, used in order to
+    /// evict keys by their value
+    key: []FileIdFileOffset,
+
+    /// frequency for the S3_FIFO
+    /// Yes, really, only 0, 1, 2, 3.
+    /// Wonder if it would be faster to pack this.
+    freq: []u2,
+
+    /// which S3_FIFO queue this frame exists in
+    in_queue: []InQueue,
+
+    /// 0..=512
+    size: []FrameOffset,
+
+    fn init(allocator: std.mem.Allocator, num_frames: usize) !FramesMetadata {
+        const rc = try allocator.alignedAlloc(
+            sig.sync.ReferenceCounter,
+            std.mem.page_size,
```

Comment:
any specific reason why alignedAlloc vs just alloc ?

Reply:
Not necessary. Just wanted to convey that they're very large buffers, tempted to just call page_allocator on them.